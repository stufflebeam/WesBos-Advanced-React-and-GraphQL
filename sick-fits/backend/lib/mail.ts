import { createTransport, getTestMessageUrl } from 'nodemailer';

const transport = createTransport({
    host: process.env.ETHEREAL_MAIL_HOST || process.env.MAIL_HOST,
    port: process.env.ETHEREAL_MAIL_PORT || process.env.MAIL_PORT,
    auth: {
        user: process.env.ETHEREAL_MAIL_USER || process.env.MAIL_USER,
        pass: process.env.ETHEREAL_MAIL_PASS || process.env.MAIL_PASS,
    },
});

const frontEndURL =
    process.env.SICKFITS_FRONTEND_URL || process.env.FRONTEND_URL;

function makeANiceEmail(text: string): string {
    return `
    <div className="email" style="
        border: 1px solid black;
        padding: 20px;
        font-family: sans-serif;
        line-height: 2;
        font-size: 20px;
    ">
        <h2>Hello There!</h2>
        <p>${text}</p>

        <p>ðŸ˜˜, Sickfits</p>
    </div>
    `;
}

export interface MailResponse {
    accepted?: string[] | null;
    rejected?: null[] | null;
    envelopeTime: number;
    messageTime: number;
    messageSize: number;
    response: string;
    envelope: Envelope;
    messageId: string;
}
export interface Envelope {
    from: string;
    to?: string[] | null;
}

export async function sendPasswordResetEmail(
    resetToken: string,
    to: string
): Promise<void> {
    // email the user a token/link to reset their password
    const info = (await transport.sendMail({
        to,
        from: 'Sickfits <test@example.com>',
        subject: 'Reset your password on Sickfits',
        html: makeANiceEmail(`Please use the following link to reset your password:<br/>
        <a href="${frontEndURL}/reset?token=${resetToken}">Reset Password</a><br/>
        <br/>
        If, for some reason, you cannot click the link, copy and paste the following into your browser:<br/>
        ${frontEndURL}/reset?token=${resetToken} <br/>
        If you did not request this, please ignore this email and your password will remain unchanged.
        `),
    })) as MailResponse;

    // console.log('[mail] info:', info);
    // console.log('[mail] Message sent: %s', info.messageId);

    // Of course, being that we are currently using an environment variable called
    // ETHEREAL_MAIL_USER, this conditional is not really necessary. But, it's good
    // to have it in here for when I build in more robust environment switching.
    if (process.env.ETHEREAL_MAIL_USER.includes('ethereal.email')) {
        console.log(
            '[mail] ðŸ“§ Message Sent! Preview URL: %s',
            getTestMessageUrl(info)
        );
    }
}
