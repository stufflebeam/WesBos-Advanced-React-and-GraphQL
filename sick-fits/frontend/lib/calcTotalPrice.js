export default function calcTotalPrice(cart) {
  //   console.log('[calcTotalPrice] cart', cart);
  if (!cart) return 0;
  return cart.reduce((tally, cartItem) => {
    // We run the following check because it is possible for an item to
    // be deleted while it is still in a user's cart.
    if (!cartItem.product) return tally;
    return tally + cartItem.product.price * cartItem.quantity;
  }, 0);
}
