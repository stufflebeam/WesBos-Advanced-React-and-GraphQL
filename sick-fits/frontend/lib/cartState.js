import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // This is our own custom provider
  // We will store data (state) and functionality (updaters) here
  // and anyone can access this data and the updaters via the consumer
  //   const cartOpen = true;
  const [cartOpen, setCartOpen] = useState(true);

  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  function openCart() {
    setCartOpen(true);
  }

  function closeCart() {
    setCartOpen(false);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, openCart, closeCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

// This is a custom hook for accessing the data in our provider (the cart local state)
function useCart() {
  // we use a consumer here, to access the local state
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, LocalStateContext, useCart };
