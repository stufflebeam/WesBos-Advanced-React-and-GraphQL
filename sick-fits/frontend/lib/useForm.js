import { useEffect, useState } from 'react';

export default function useForm(initial = {}) {
  // create a state object for our form inputs
  const [inputs, setInputs] = useState(initial);
  // const initialValues = Object.values(initial).join('');
  // This 'feels' better than the suggested solution (above), as it avoids useEffect()
  // being called due to object property order not being guaranteed.
  // This way, it only gets called when the form is first populated with fetched data
  // and the values go from '' to something else.
  const initialValuesEmpty = Object.values(initial).join('') === '';
  // console.log('[useForm] initialValuesEmpty: ', initialValuesEmpty);

  useEffect(() => {
    // This function runs when the things we are watching change
    // console.log(
    //   '[useForm] useEffect() initialValuesEmpty: ',
    //   initialValuesEmpty
    // );
    setInputs(initial);
  }, [initialValuesEmpty]);

  // create a function that updates the state object
  const handleChange = (e) => {
    let { value, name, type } = e.target;

    // check to see if the input is a number, and if it is, convert it to a number
    if (type === 'number') {
      value = parseInt(value);
    }

    // check to see if the input is a file and handle it accordingly if so
    if (type === 'file') {
      [value] = e.target.files;
    }

    // console.log('[handleChange]', { value, name, type });
    setInputs({
      // copy the existing state first
      ...inputs,
      // update the state with the new input value
      [name]: value,
    });
  };

  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    // setInputs({}); // doesn't work
    const blankState = Object.fromEntries(
      Object.entries(inputs).map(([key, value]) => [key, ''])
    );
    setInputs(blankState);
  }

  // return the things we wan to surface from this custom hook
  // specifically, an object that includes the state and a function to update it
  return { inputs, handleChange, resetForm, clearForm };
}
