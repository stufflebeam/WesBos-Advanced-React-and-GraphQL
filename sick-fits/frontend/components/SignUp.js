import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import styled from 'styled-components';
import useForm from '../lib/useForm';
import DisplayError from './ErrorMessage';
import Form from './styles/Form';
import { CURRENT_USER_QUERY } from './User';

const AccountCreatedNotificationStyles = styled.p`
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const SIGNUP_MUTATION = gql`
  mutation SIGNUP_MUTATION(
    $email: String!
    $name: String!
    $password: String!
  ) {
    createUser(data: { email: $email, name: $name, password: $password }) {
      id
      email
      name
    }
  }
`;

export default function SignUp() {
  const { inputs, handleChange, resetForm, clearForm } = useForm({
    email: '',
    name: '',
    password: '',
  });

  const [signUp, { data, loading, error }] = useMutation(SIGNUP_MUTATION, {
    variables: inputs,
    // refetch the currently logged in user
    // This doesn't apply here because signing up doesn't log the user in
    // So, they'll still go through the login process and the CURRENT_USER_QUERY
    // will be run as part of that.
    // refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });

  // console.log({ data, loading, error });

  async function handleSubmit(e) {
    e.preventDefault();
    // console.log('[SignUp] handleSubmit() inputs:', inputs);
    // Send the email and password to the GraphQL API
    const res = await signUp().catch(console.error);
    // console.log('[SignUp] handleSubmit() res:', res);
    // console.log('[SignUp] handleSubmit()', { data, loading, error });
    resetForm();
  }

  // const authError =
  //   data?.authenticateUserWithPassword?.__typename ===
  //   'UserAuthenticationWithPasswordFailure'
  //     ? data?.authenticateUserWithPassword
  //     : undefined;

  if (data?.createUser) {
    return (
      <AccountCreatedNotificationStyles>
        Account created for {data.createUser.email}! <br />
        Please go ahead and Sign In!
      </AccountCreatedNotificationStyles>
    );
  }

  return (
    // Using method="post" to prevent the possibility of a user's password being passed as a query parameter
    // in the browser's address bar.
    <Form method="POST" onSubmit={handleSubmit}>
      <h2>Sign Up for an Account</h2>
      <DisplayError error={error} />
      {/* {console.log(
        '[SignUp] data?.authenticateUserWithPassword?.message:',
        data?.authenticateUserWithPassword?.message
      )} */}
      {/* <fieldset disabled={loading} aria-busy={loading}> */}
      <fieldset>
        {/* Optionally switch the 'Account Created' message's location to here, if you want
            the sign-up form to remain visible after account creation. */}
        {/* {data?.createUser && (
          <AccountCreatedNotificationStyles>
            Account created for {data.createUser.email}! <br />
            Please go ahead and Sign In!
          </AccountCreatedNotificationStyles>
        )} */}
        <label htmlFor="name">
          Your Name
          <input
            type="text"
            name="name"
            id="name"
            placeholder="Your Name"
            autoComplete="name"
            onChange={handleChange}
            value={inputs.name || ''}
          />
        </label>
        <label htmlFor="email">
          Email
          <input
            type="email"
            name="email"
            id="email"
            placeholder="Your email address"
            autoComplete="email"
            onChange={handleChange}
            value={inputs.email || ''}
          />
        </label>
        <label htmlFor="password">
          Password
          <input
            type="password"
            name="password"
            id="password"
            placeholder="Password"
            autoComplete="current-password"
            onChange={handleChange}
            value={inputs.password || ''}
          />
        </label>
        <button type="submit">Sign Up</button>
      </fieldset>
    </Form>
  );
}
