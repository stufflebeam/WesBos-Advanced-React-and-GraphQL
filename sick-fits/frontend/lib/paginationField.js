// import {PAGINATION_QUERY} from '../graphql/queries'; // TODO: move all queries to graphql folder
import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false, // This tells Apollo that we will take care of everything.
    read(existing = [], { args, cache }) {
      //   console.log({ existing, args, cache });
      const { skip, first } = args;
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      //   console.log({ first, skip, data });
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1; // Math.ceil(skip / first) + 1;
      const pageCount = Math.ceil(count / first);
      console.log('[paginationField] read():', {
        skip,
        first,
        count,
        page,
        pageCount,
      });

      // Check to see if we have the requested items in the cache.
      const items = existing.slice(skip, skip + first).filter((x) => x); // TODO: Should this be .filter((x) => !!x);

      /* 
      IF
        There are items
        AND the number of items is not enough to satisfy the request (i.e. there are not enough to fill the page)
        AND we are on the last page
      THEN just send the data back to Apollo
      */

      // TODO: Handle the case where a user tries to navigate to a page that doesn't exist.
      //       For example, if there are 10 items total, and we're showing 4 items per page,
      //       and the user navigates to page 5, we should either route them to the main
      //       products page OR route them to the last products page by updating the URL
      //       from products/5 to products/3 and showing them the last 2 items.

      if (items.length && items.length !== first && page === pageCount) {
        return items;
      }

      if (items.length !== first) {
        // If we don't have all the items, we need to fetch them.
        return false;
      }

      // If we have all the items, we can just return them.
      if (items.length) {
        console.log(
          `[paginationField]: There are ${items.length} items in the cache. Sending them to Apollo...`
        );
        return items;
      }

      // If either of the previous return statements didn't return anything, we need to fetch the items,
      // so we return false.
      return false;

      // When we run a query, the first thing that Apollo does is ask the read() function for those items.
      // When this happens, we can do one of two things:
      // 1. If we have the data, we can return it.
      // 2. If we don't have the data, we can return null or false, which will tell Apollo to run the query via the network.
    },
    merge(existing, incoming, { args, cache }) {
      // When Apollo gets query results back from the network for items that the read() function returned false for,
      // it will call the merge() function to merge the results into the cache.
      // So, this is where we decide how we want to put the returned data into the cache.
      const { skip, first } = args;

      console.log('[paginationField]: merging items from the network...]', {
        existing,
        incoming,
        args,
        cache,
      });
      const merged = existing ? existing.slice(0) : [];

      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }
      console.log('[paginationField] merge():', { merged });

      // return the merged items
      // at this point, Apollo will call the read() function again to see if it can return the
      // merged items from the cache.
      return merged;
    },
  };
}
