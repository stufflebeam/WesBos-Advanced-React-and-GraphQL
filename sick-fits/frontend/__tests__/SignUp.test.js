import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import { CartStateProvider } from '../lib/cartState';
import SignUp, { SIGNUP_MUTATION } from '../components/SignUp';
// import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser } from '../lib/testUtils';

const me = fakeUser();
const password = '12345678';

const mocks = [
  // Mutation Mock
  {
    request: {
      query: SIGNUP_MUTATION,
      variables: {
        name: me.name,
        email: me.email,
        password,
      },
    },
    result: {
      data: {
        createUser: {
          __typename: 'User',
          id: 'abc123',
          name: me.name,
          email: me.email,
        },
      },
    },
  },
  // Current User Query Mock
  //   {
  //     request: { query: CURRENT_USER_QUERY },
  //     result: { data: { authenticatedItem: me } },
  //   },
];

describe('<SignUp />', () => {
  it('renders and matches snapshot', () => {
    const { container, debug } = render(
      <MockedProvider>
        <SignUp />
      </MockedProvider>
    );
    // debug(container);
    expect(container).toMatchSnapshot();
  });

  it('calls the mutation and notifies the user of successful account creation properly', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <SignUp />
      </MockedProvider>
    );

    // Type into the boxes in the form
    await userEvent.type(screen.getByPlaceholderText('Your Name'), me.name);
    await userEvent.type(
      screen.getByPlaceholderText('Your email address'),
      me.email
    );
    // TODO: Add tests for insufficiently-complex passwords
    await userEvent.type(screen.getByPlaceholderText('Password'), password);
    // Click the Submit button
    await userEvent.click(screen.getByText('Sign Up'));

    // Wait for the mutation to resolve
    await screen.findByText(
      `Account created for ${me.email}! Please go ahead and Sign In!`
    );

    // debug(container);

    expect(container).toMatchSnapshot();
  });
});
