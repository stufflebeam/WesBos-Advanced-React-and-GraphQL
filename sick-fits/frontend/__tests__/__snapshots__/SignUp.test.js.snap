// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> calls the mutation and notifies the user of successful account creation properly 1`] = `
<div>
  <p
    class="SignUp__AccountCreatedNotificationStyles-sc-1yuwtdx-0 eoxVZf"
  >
    Account created for 
    Delmer.Smith@yahoo.com
    ! 
    <br />
    Please go ahead and Sign In!
  </p>
</div>
`;

exports[`<SignUp /> renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 hVEEdl"
    method="POST"
  >
    <h2>
      Sign Up for an Account
    </h2>
    <fieldset>
      <label
        for="name"
      >
        Your Name
        <input
          autocomplete="name"
          id="name"
          name="name"
          placeholder="Your Name"
          type="text"
          value=""
        />
      </label>
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          id="email"
          name="email"
          placeholder="Your email address"
          type="email"
          value=""
        />
      </label>
      <label
        for="password"
      >
        Password
        <input
          autocomplete="current-password"
          id="password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Sign Up
      </button>
    </fieldset>
  </form>
</div>
`;
